[test] Fix storage errors for unit testing

m-Peter opened the pull request:
Work towards https://github.com/onflow/developer-grants/issues/216

## Description

In `InterpreterConfig.ContractValueHandler` function, we would override the interpreter's shared state with a new storage based on emulator's ledger.
```go
storage := runtime.NewStorage(
	r.backend.blockchain.NewScriptEnvironment(),
	nil,
)
// Update the storage to reflect the changes
// from deployments in setup() function.
inter.SharedState.Config.Storage = storage
```
This proved to be problematic, as the new storage would not contain the constructed composite values needed for testing, e.g. `I.Test`. Resulting in the following error:
```shell
error: internal error: slab (0x0.7) not found: slab not found for stored value
```
The `slab (0x0.7)` would contain the emulator backend:
```cadence
access(all) contract Test {

    /// backend emulates a real network.
    ///
    access(self) let backend: AnyStruct{BlockchainBackend}

    init(backend: AnyStruct{BlockchainBackend}) {
        self.backend = backend
    }
    ...
}
```
To fix this, we simply store every slab from the new storage, to the current environment's storage.

______

<!-- Complete: -->

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence-lint/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

turbolent commented on test/test_runner.go:
```diff
@@ -581,14 +584,14 @@ func (r *TestRunner) interpreterContractValueHandler(
+				// All contracts are deployed on EmulatorBackend's
+				// blockchain, so we construct a storage based on
+				// its ledger.
+				blockchainStorage := runtime.NewStorage(
```
Is it maybe possible to just use the same storage, instead of creating a new one and copying over the slabs?

m-Peter replied:
I guess it has to do with the fact that the special `setup()` function is the one that makes the deployments.

m-Peter replied:
> I'll see if there is an easy way to re-use the storage.

If this proves to require more elaborate work, could we consider leaving this as a technical debt? Releasing this will provide better DevEx :pray: 

```cadence
let randomSource = scriptResult.returnValue! as! RandomBeaconHistory.RandomSource
let expectedRandomSource = RandomBeaconHistory.RandomSource(
    blockHeight: getCurrentBlockHeight() - 1,
    value: [0, 1, 1, 2, 3, 5, 8]
)
Test.assertEqual(expectedRandomSource, randomSource)
```
The above use of `RandomBeaconHistory.RandomSource(...)` would trigger the error in the PR's description.

turbolent replied:
@m-Peter What problems are you running into with the original code?

m-Peter replied:
In certain cases, when creating values of types defined in imported contracts, e.g.

```cadence
import Test
import "RandomBeaconHistory"

// ... contracts deployment & setup

let randomSource = scriptResult.returnValue! as! RandomBeaconHistory.RandomSource
let expectedRandomSource = RandomBeaconHistory.RandomSource(
    blockHeight: getCurrentBlockHeight() - 1,
    value: [0, 1, 1, 2, 3, 5, 8]
)
Test.assertEqual(expectedRandomSource, randomSource)
```

we are no longer able to call some methods on the `Test` contract. I get the error:
```shell
error: internal error: slab (0x0.7) not found: slab not found for stored value
```
This comes from this line: https://github.com/onflow/cadence-tools/blob/master/test/test_runner.go#L588-L590
We inject a different storage instance in the `inter`, which does not contain the slabs allocated for the `Test` contract.

bluesign replied:
Just committing storage should be enough. https://github.com/onflow/cadence/blob/f5a7b6e3c84ca9757792018eb5a821f79b569cd5/runtime/storage.go#L230C18-L230C25 no need to set new storage

turbolent replied:
Maybe the problem is a missing commit of the data?

m-Peter replied:
I will try it out :pray:

m-Peter replied:
@bluesign You mean this kind of context?
```go
ctx := runtime.Context{
	Interface:   r.backend.blockchain.NewScriptEnvironment(),
	Location:    testScriptLocation,
	Environment: env,
}
```

bluesign replied:
I checked it out in detail, unfortunately not many options ( I am not even sure there is 100% correct solution ) 

Situation is `Test` is running in script context, but deployContract etc running transaction. One option for simple contracts can be syncing storage after deploy contract phase. ( basically getting Contract Composite from emulator storage, Clone(), and write to interpreter storage ) Even this passes current tests, unfortunately this doesn't work, if contract makes changes to somewhere else. ( for example creating another account, or saving something to account storage )

bluesign replied:
@turbolent commit was my main thought too, but deploy is running a transaction, ofc we can always change deploy to deploy in the current context.

SupunS replied:
I agree, had a look as well, and it is not straightforward at this point to 'share' the storage for the test-script and the tested code (deployed contract / transaction).

I think we can merge this as-is and unblock developers, and then look into a long-term solution separately.
Given this will be used only during tests, I don't think the overhead of copying would be a deal-breaker for now.

turbolent replied:
Sounds good! Can we please create a follow-up issue (if it does not exist yet)?

turbolent reviewed (commented):
Good find and tests 👍

bluesign commented on test/test_runner.go:
```diff
@@ -599,6 +602,30 @@ func (r *TestRunner) interpreterContractValueHandler(
+					if err != nil {
+						panic(err)
+					}
+					err = storage.Commit(inter, true)
+					if err != nil {
+						panic(err)
+					}
```
I think this can go outside for (not sure though)

m-Peter replied:
It actually makes no difference whether `storage.Commit(inter, true)` is called or not :sweat_smile:
I am quite unsure as to why this is the case. But anyway, I have moved it out of the `for` loop :+1: 
Updated in https://github.com/onflow/cadence-tools/pull/226/commits/ec94c9fcf452697136cb6f6fe6f76e14d3216d07

m-Peter commented:
@SupunS I have rebased this on top of master and moved the `storage.Commit` outside the `for` fool (https://github.com/onflow/cadence-tools/commit/ec94c9fcf452697136cb6f6fe6f76e14d3216d07).
If there's no current alternative to overcome the issue in the description, could we consider getting this in, and creating a tech debt issue or something similar.

turbolent reviewed (approved):
👌
